---
title: "Problemset 4"
subtitle: "ME.440.825 Quantitative Neurogenomics"
author: "<Your Name Here>"
date: ''
output:
  html_document:
    toc: yes
    toc_level: 3
    toc_float: yes
    theme: yeti
    df_print: paged
  pdf_document: default
editor_options:
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Part 1: RNA-seq Pseudoalignment

For this problem set, you will perform pseudoalignment on RNA-seq data from S. Cerevisiae (baker's yeast), which has a conveniently small genome. Yeast has been used as a model to study protein aggregation in neurodegenerative disorders. In Chen et al. 2020 (https://doi.org/10.1038/s41467-020-14525-4), the authors use a yeast model of consitutive amyloid-$\beta$42 (A$\beta$42), which is more prone to aggregation than the more common form, A$\beta$40. They find that flavin mononucleotide (FMN) supplementation reduces A$\beta$ toxicity, and use RNA-seq to investigate whether FMN supplementation causes transcriptional changes. 

### Task 1
Download FASTQ files containing raw sequencing data from the paper. Also download the corresponding metadata.
GEO series number: GSE128905.
For time's sake, just download the files corresponding to 4 samples (1 per condition):
GSM3688093: Control
GSM3688105: Control + FMN
GSM3688117: A$\beta$42
GSM3688129: A$\beta$42 + FMN
```{bash getData}

# Create accession variable and directory structure
GEO="GSE128905"
mkdir -p data/raw/${GEO}
cd data/raw/${GEO}

# Get .json file with metadata for all samples
# This operation sometimes fails with errors because of rate limiting at NCBI.  If it fails, just try again.
#ffq $GEO > $GEO.json

# Use jq to extract the ftp links for the fastq files from the .json and pass to wget to download (4 at a time)
# Note: this will download all fastq files, and will be approximately 60 Gb (24 * 2.5Gb) of raw data. Please be sure you have sufficient disk space available.
jq -r '.[].geo_samples | .["GSM3688093","GSM3688105","GSM3688117","GSM3688129"].samples | .[].experiments | .[].runs | .[].files.ftp | .[].url ' $GEO.json | xargs -n 1 -P 4 wget -c

# Use jq to extract the metadata for each sample and write to a .tsv file
jq -r '.[].geo_samples | .["GSM3688093","GSM3688105","GSM3688117","GSM3688129"].samples | .[] | [.accession, (.experiments | .[].runs | .[] | .files.ftp | .[] | .[]), (.attributes | .[])] | @tsv' $GEO.json > $GEO.tsv

```

### Task 2
How many reads are in each of the first two fastq files in your directory? Print the names of the files along with the numbers of reads. Do the two files have the same number of reads? Why or why not?
```{bash numReads}
GEO="GSE128905"
cd data/raw/${GEO}
echo $(ls *.fastq.gz | head -n 1) 
numLines=$(gunzip -c $(ls *.fastq.gz | head -n 1) | wc -l)
echo $(($numLines/4)) reads
echo $(ls *.fastq.gz | head -n 2 | tail -n 1)
numLines=$(gunzip -c $(ls *.fastq.gz | head -n 2 | tail -n 1) | wc -l)
echo $(($numLines/4)) reads
```
They have the same number of reads because these two fastq files contain paired-end reads.

###Task 3

```{bash qc}
echo $(ls *.fastq.gz) | xargs -n 1 -P 4 fastqc
```



```{bash}

```



```{bash}
 
```



```{bash}

```



```{bash}

```



```{bash}

```



```{bash}

```

## Part 2: 

```{r}

```


## PSet Submission

When you have completed the above exercises, please make sure that you have added your name and date into the yaml header at the top and "knit" this document.

![](images/paste-70D156D7.png)

When you are "knitting" please make sure that you set the knit directory to: "Document Directory" to that RStudio knows where to find all of the accessory files associated with this problemset.

![](images/paste-5760B0D6.png)

Once this completes without error, please upload the compiled .html document to the "Problem Set 1" assignment on Canvas.

## Session Information
```{r}
sessionInfo()
```
