---
title: "Problemset 4"
subtitle: "ME.440.825 Quantitative Neurogenomics"
author: "<Your Name Here>"
date: ''
output:
  html_document:
    toc: yes
    toc_level: 3
    toc_float: yes
    theme: yeti
    df_print: paged
  pdf_document: default
editor_options:
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

For this problem set, you will perform pseudoalignment on RNA-seq data from S. Cerevisiae (baker's yeast), which has a conveniently small genome. Yeast has been used as a model to study protein aggregation in neurodegenerative disorders. In Chen et al. 2020 (https://doi.org/10.1038/s41467-020-14525-4), the authors use a yeast model of consitutive amyloid-$\beta$42 (A$\beta$42), which is more prone to aggregation than the more common form, A$\beta$40. They find that flavin mononucleotide (FMN) supplementation reduces A$\beta$ toxicity, and use RNA-seq to investigate whether FMN supplementation causes transcriptional changes. 

### Task 1
Ensure that you have installed all of the necessary tools in your conda environment:
``` {bash}
#conda install -c bioconda fastqc kallisto igv
#conda install jq
#pip install ffq gget multiqc
```

### Task 2
Download FASTQ files containing raw sequencing data from the paper. Also download the corresponding metadata and write it to a .tsv file.
GEO series number: GSE128905.
For time's sake, just download the files corresponding to 4 out of the 48 total samples (1 per condition):
GSM3688093: Control
GSM3688105: Control + FMN
GSM3688117: A$\beta$42
GSM3688129: A$\beta$42 + FMN

```{bash getData}

# Create accession variable and directory structure
GEO="GSE128905"
mkdir -p data/raw/${GEO}
cd data/raw/${GEO}

# Get .json file with metadata for all samples
# This operation sometimes fails with errors because of rate limiting at NCBI.  If it fails, just try again.
ffq $GEO > $GEO.json

# Use jq to extract the ftp links for the fastq files for your desired samples from the .json and pass to wget to download (4 at a time)
jq -r '.[].geo_samples | .["GSM3688093","GSM3688105","GSM3688117","GSM3688129"].samples | .[].experiments | .[].runs | .[].files.ftp | .[].url ' $GEO.json | xargs -n 1 -P 4 wget -c

# Use jq to extract the metadata for each sample and write to a .tsv file
jq -r '.[].geo_samples | .["GSM3688093","GSM3688105","GSM3688117","GSM3688129"].samples | .[] | [.accession, (.experiments | .[].runs | .[] | .files.ftp | .[] | .[]), (.attributes | .[])] | @tsv' $GEO.json > $GEO.tsv

```

### Question 2.1
In your raw data directory, you should have 8 FASTQ files, where four filenames end with "_1.fastq.gz" and the other four end with "_2.fastq.gz. Why are there two files for each SRR ID number?


### Task 3
Download and expand the pre-built S. Cerevisiae kallisto index, which you will use for alignment.
```{bash}
GEO="GSE128905"
mkdir -p metadata/$GEO/kallisto_index
cd metadata/$GEO/kallisto_index
IDX_URL="https://github.com/pachterlab/kallisto-transcriptome-indices/releases/download/ensembl-96/saccharomyces_cerevisiae.tar.gz"
wget -c $IDX_URL
tar -xzvf saccharomyces_cerevisiae.tar.gz # Expands into a new directory called saccharomyces_cerevisiae
```

### Question 3.1 
What are some advantages to using Kallisto here over other alignment methods?

### Question 3.2
This dataset includes RNA-seq data from yeast where A$\beta$42 has been introduced via transformation. Given the pre-built index that you have downloaded, will you be able to map reads to the A$\beta$42 sequence? Why or why not?

### Question 3.3
This is an RNA-sequencing data set. What if you had an ATAC-seq dataset? Which aligner might you choose and why? 

### Task 4
Run kallisto pseudoalignments on your 4 samples. You will not need to visualize your mapping in IGV (so choose your run options accordingly). 
```{bash}
# Build dir structure for results
mkdir -p results/$GEO/kallisto
cd results/$GEO/kallisto

# Run kallisto pseudoalignments
for fastq in $(ls ../../../data/raw/$GEO/*1.fastq.gz); 
    do
        fastq_pair=../../../data/raw/$GEO/$(basename $fastq 1.fastq.gz)2.fastq.gz 
        SAMPLE_NAME=$(basename $fastq _1.fastq.gz);
        echo "Performing kalliso pseudoalignment for $SAMPLE_NAME";
        mkdir $SAMPLE_NAME
        time kallisto quant -i ../../../metadata/$GEO/kallisto_index/saccharomyces_cerevisiae/transcriptome.idx -o $SAMPLE_NAME $fastq $fastq_pair 2>$SAMPLE_NAME/$SAMPLE_NAME.log;
    done 
```

### Question 4.1
Question about what the arguments correspond to for Kallisto and Bowtie calls

### Task 5
Use multiqc to generate a project QC report in your working directory. You will use this report to answer the questions that follow.
```{bash}
multiqc .
```

### Question 5.1
Questions about the QC documents.


## PSet Submission

When you have completed the above exercises, please make sure that you have added your name and date into the yaml header at the top and "knit" this document.

![](images/paste-70D156D7.png)

When you are "knitting" please make sure that you set the knit directory to: "Document Directory" to that RStudio knows where to find all of the accessory files associated with this problemset.

![](images/paste-5760B0D6.png)

Once this completes without error, please upload the compiled .html document to the "Problem Set 1" assignment on Canvas.

## Session Information
```{r}
sessionInfo()
```
