---
title: 'Differential expression analysis using DESeq2'
output: html_document
date: 10/14/2022
---

```{r}
library(DESeq2)
library(ggplot2)
library(ggrepel)
library(pheatmap)
```

# 1. Load data

RNA-seq results
```{r}
load("./Results/results_count.Rda")
ls()
```

Sample sheet
```{r}
sampleSheet <- read.table("./Sample/SampleSheet.csv", sep=",", header=T, stringsAsFactors = F)

# factorize
sampleSheet$Region <- factor(sampleSheet$Region, levels=c("Dorsal", "Ventral"))
sampleSheet$Replicate <- factor(sampleSheet$Replicate, levels=c("Rep1", "Rep2", "Rep3"))
```

# 2. EDA

### Generate a count matrix
```{r}
fCounts.count <- fCounts[, sampleSheet$Sid]
rownames(fCounts.count) <- fCounts$gene_id
summary(fCounts.count)
```

## TPM
```{r}
tpm <- fCounts.count/(fCounts$length/1000)
tpm <- mapply("/", tpm, colSums(tpm)/10^6)
rownames(tpm) <- fCounts$gene_id
colSums(tpm) # should be 10^6 across all the samples.
```

### PCA
```{r}
log10tpm <- log10(tpm+1)
log10tpm.var.idx <- which(apply(log10tpm, 1, var)>0)

pca.obj <- prcomp(t(log10tpm[log10tpm.var.idx,]), center=T, scale=T)
pca.res <- data.frame(sampleSheet, PC1=pca.obj$x[,1], PC2=pca.obj$x[,2])
p <- ggplot(pca.res, aes(x=PC1, y=PC2, color=Region, shape=Replicate))
p <- p + geom_point(size=3)
p + ggtitle("PCA")
```

### Heatmap using pheatmap
```{r}
# sample annotation
ht_annot <- data.frame(row.names=sampleSheet$Sid, Region=sampleSheet$Region, Replicate=sampleSheet$Replicate)

# sample annotation color
ht_annot_color <- list(Region=setNames(c("red","black"), c("Dorsal","Ventral")),
                       Replicate=setNames(c("blue","green","orange"), c("Rep1","Rep2","Rep3")))

pheatmap(log10tpm[log10tpm.var.idx,], 
         scale = "row",
         annotation_col=ht_annot, 
         annotation_colors=ht_annot_color,
         show_rownames = F)
```

# 3. DESeq2

### Generate a DESeq object
```{r}
dds <- DESeqDataSetFromMatrix(fCounts.count, colData=sampleSheet, design= ~ Replicate + Region)
dds
```

### Normalization and model fitting
```{r}
dds <- DESeq(dds)

# Size factors
sizeFactors(dds)

# Dispersion estimates
plotDispEsts(dds)
```
* Size factors are generally around 1.  
* Dispersion plot: The fitted curve should be surrounded by gene-est. Also, check if dispersion decreases with mean expression increasing. 

### Hypothesis test
```{r}
res <- results(dds, contrast = c("Region", "Dorsal", "Ventral")) # Choose the column in the sample sheet to compare the samples. The last element is the reference level in the comparison.
res
hist(res$pvalue) 
```
* The distribution of p-value should be either flat or enriched with lower p-value.

### MA plot
```{r}
DESeq2::plotMA(res)
```

# 4. Differential expression analysis

### Summary
```{r}
de.res <- gtf[, c("gene_id", "gene_name", "gene_type")]

# TPM group mean
de.res$tpm.dorsal <- rowMeans(tpm[, subset(sampleSheet, Region=="Dorsal")$Sid])
de.res$tpm.ventral <- rowMeans(tpm[, subset(sampleSheet, Region=="Ventral")$Sid])

# DESeq2 results
de.res$log2fc <- res$log2FoldChange
de.res$pval <- res$pvalue
de.res$padj <- res$padj
```

### Significant genes
```{r}
de.res$sig <- NA # default
de.res$sig[!is.na(de.res$padj)] <- "Insig"
de.res$sig[which(de.res$padj<0.01)] <- "Sig"
de.res$sig[which(de.res$padj<0.01 & abs(de.res$log2fc)>1)] <- "SigFC"
table(de.res$sig)

de.res.sigFC <- subset(de.res, sig=="SigFC") # just for reference
```

### Genome-wide scatter plot
```{r}
# x axis: average TPM in Dorsal
# y axis: average TPM in Ventral

# find the 5 top significant genes in both up- and down-regulated.
temp <- de.res
down.idx <- match(de.res.sigFC$gene_id[order(de.res.sigFC$log2f)[1:5]], de.res$gene_id)
up.idx <- match(de.res.sigFC$gene_id[order(de.res.sigFC$log2f, decreasing = T)[1:5]], de.res$gene_id)
temp$label <- NA
lab.idx <- c(down.idx, up.idx)
temp$label[lab.idx] <- temp$gene_name[lab.idx]

p <- ggplot(temp, aes(x=log10(tpm.dorsal), y=log10(tpm.ventral), label=label))
p <- p + geom_point(data=subset(temp, sig == "Insig"), col="gray60", alpha=0.3)
p <- p + geom_point(data=subset(temp, sig != "Insig"), aes(col=sig), alpha=0.7)
p <- p + scale_color_manual(values=c("blue","red"), label=c("pval<0.01", "pval<0.01 & |FC|>2"))
p <- p + geom_text_repel()
p <- p + geom_abline(slope = 1, intercept = 0, linetype="dashed")
p <- p + xlab("log10(TPM) Mean in Dorsal") + ylab("log10(TPM) Mean in Ventral")
p
```

### Volcano plot
```{r}
temp <- subset(de.res, !is.na(sig))
p <- ggplot(temp, aes(x=log2fc, y=-log10(padj), col=sig))
p <- p + geom_point(alpha=0.7)
p <- p + scale_color_manual(values=c("gray", "blue", "red"), label=c("Insig.", "pval<0.01", "pval<0.01 & |FC|>2"))
p <- p + xlab("Log2 Fold Change")
p <- p + ylab("-Log10 Adjusted P-value")
p
```

# 5. GO

```{r}
library(clusterProfiler)
# options(connectionObserver = NULL) # run this line in case of connection error below.
library(org.Mm.eg.db)
```

### Remove suffix from gene id to make gene id compatible with GO database.
```{r}
temp <- strsplit(de.res$gene_id, split="[.]")
length(unique(sapply(temp, "[[", 1)))==length(temp) # check if removing suffix does not affect uniqueness of gene id.
de.res$gene_id2 <- sapply(temp, "[[", 1)
```

### Enrichment analysis
```{r}
ego <- enrichGO(gene = subset(de.res, sig=="Sig")$gene_id2, 
                universe = subset(de.res, !is.na(sig))$gene_id2,
                keyType = "ENSEMBL",
                OrgDb = org.Mm.eg.db, 
                ont = "BP", 
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)
dotplot(ego, showCategory=50)

ego.up <- enrichGO(gene = subset(de.res, sig=="Sig" & log2fc>0)$gene_id2, 
                universe = subset(de.res, !is.na(sig))$gene_id2,
                keyType = "ENSEMBL",
                OrgDb = org.Mm.eg.db, 
                ont = "BP", 
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)
dotplot(ego.up, showCategory=50)

ego.down <- enrichGO(gene = subset(de.res, sig=="Sig" & log2fc<0)$gene_id2, 
                universe = subset(de.res, !is.na(sig))$gene_id2,
                keyType = "ENSEMBL",
                OrgDb = org.Mm.eg.db, 
                ont = "BP", 
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)
dotplot(ego.down, showCategory=50)

#write.csv(ego, "./Results/GO_results.csv", quote=F)
```

# 6. GSEA
```{r}
# Make an ordered and named vector of fold changes
temp <- subset(de.res, !is.na(log2fc))
temp <- temp[order(temp$log2fc, decreasing = T),]
temp <- setNames(temp$log2fc, temp$gene_id2)

gsea.res <- gseGO(geneList = temp, 
                  OrgDb = org.Mm.eg.db, 
                  ont = "BP", 
                  keyType = "ENSEMBL",
                  pvalueCutoff = 0.1)

#gseaplot(gsea.res, geneSetID = 'GO:0015850')

# Write results to file
#write.csv(gsea.res@result, "./Results/GSEA_results.csv", quote=F)
```

```{r}
sessionInfo()
```

